java/opendcs/src/main/java/decodes/cwms/algo/InflowEstimationAlgo.java

```diff
@@ -0,0 +1,518 @@
+		long durationMs = IntervalCodes.getIntervalSeconds(durationPeriod) * 1000L;
+		long windowStart = windowEnd - durationMs;
+		//Calculate as end-of-period stepped
+		Long s = values.ceilingKey(windowStart);
+		while(s != null)
+		{
+			Long prev = values.lowerKey(s);
+			long segStart = Math.max(prev == null ? Long.MIN_VALUE : prev, windowStart);
+			long segEnd = Math.min(s, windowEnd);
+
+			if(segEnd > segStart)
+			{
+				double weight = (double) (segEnd - segStart) / (double) durationMs;
+				weightedSum += weight * values.get(s);
+				totalWeights += weight;
+			}
+
+			if(s >= windowEnd)
+			{
+				break;
+			}
+			s = values.higherKey(s);
+		}
+		if(totalWeights == 0.0)
+		{
+			throw new NotEnoughDataException(
+					"Not enough data to perform period average computation to interval seconds: " + durationPeriod
+			);
+		}
+		return weightedSum / totalWeights;
+	}
+
+	private void extendTimeSeries(CTimeSeries timeSeries) throws DbCompException
+	{
+		Date start = Date.from(_aggregatePeriodEnd.toInstant().minusSeconds(IntervalCodes.getIntervalSeconds(durationPeriod)));
+		if(timeSeries.findWithin(start, 0) != null && timeSeries.findWithin(_aggregatePeriodEnd, 0) != null)
+		{
+			return;
+		}
+		try
+		{
+			timeSeriesDAO.fillTimeSeries(timeSeries, start, _aggregatePeriodEnd);
+		}
+		catch(DbIoException | BadTimeSeriesException e)
+		{
+			throw new DbCompException("Could not retrieve time series: " + timeSeries.getTimeSeriesIdentifier(), e);
+		}
+	}
+
+	private double calculateHoldout() throws DbCompException, NotEnoughDataException, NoConversionException
+	{
+		NavigableMap<Long, Double> storageRaw;
+		if(stagePoolTs != null && stagePoolTs.getTimeSeriesIdentifier() != null)
+		{
+			extendTimeSeries(stagePoolTs);
+			TSUtil.convertUnits(stagePoolTs, stagePoolTs.getTimeSeriesIdentifier().getStorageUnits());
+			storageRaw = rate(conn, stagePoolTs, stageStorRatingSet);
+			debug2(_aggregatePeriodEnd + ": rated storage: " + storageRaw);
+		}
+		else
+		{
+			extendTimeSeries(storageTs);
+			TSUtil.convertUnits(storageTs, storageTs.getTimeSeriesIdentifier().getStorageUnits());
+			storageRaw = getValues(storageTs);
+			debug2(_aggregatePeriodEnd + ": storage ts: " + storageRaw);
+		}
+		//Units are in m3 - calculated to cms
+		double holdout = calculateHoldout(storageRaw);
+		debug2(_aggregatePeriodEnd + ": averaged holdout: " + holdout);
+		return holdout;
+	}
+
+	private double calculateHoldout(NavigableMap<Long, Double> values) throws NotEnoughDataException
+	{
+		double value = findInterp(values, _aggregatePeriodEnd);
+		Date previousInterval = Date.from(_aggregatePeriodEnd.toInstant()
+				.minusSeconds(IntervalCodes.getIntervalSeconds(durationPeriod)));
+		double previousVariable = findInterp(values, previousInterval);
+		return (value - previousVariable) / IntervalCodes.getIntervalSeconds(durationPeriod);
+
+	}
+
+	private static double findInterp(NavigableMap<Long, Double> values, Date time) throws NotEnoughDataException
```

Comment:
the computation driver already helps interp if input is configured for interp? why are we duplicating here?

Reply:
Can you point to me where in the computations that happens? The goal here is to allow inputs to be irregular data, and then process the average or delta based on the inflow time series' duration